### DISCUSSION TITLE
Customizing Templates

### DISCUSSION THREAD

**Discussion by sbock17** [nvim-telekasten/telekasten.nvim#101] [Ideas]
Hello, the basic idea, as @lambtho12 stated, would be to „refactor“ the telekasten plugin in terms of custom templates, to have such settings in one folder.

My first goal is to modify the „long date“ variable, `hdate`, in order to show German names for day of the week and months. 
`hdate` is defined by 

```lua
    dates.hdate = daymap[wday]
        .. ", "
        .. monthmap[dinfo.month]
        .. " "
        .. dinfo.day
        .. daysuffix(dinfo.day)
        .. ", "
        .. dinfo.year
```

These definitions of `daymap` and `monthmap` are what I am searching for:

```lua
local daymap = { "Monday", "Tuesday",  "Wednesday", "Thursday", "Friday", "Saturday", "Sunday", }
local monthmap = { "January",  "February", "March", "April", "May", "June", "July", "August", "September", "October", "November",  "December", }
```

And I changed them to
```lua
daymap = {"Montag","Dienstag","Mittwochj","Donnerstag", "Freitag","Samstag","Sonntag",},
monthmap = {"Januar","Februar","März","April","Mai","Juni","Juli","August","September","Oktober","November","Dezember",},
```

And how would the month days be calculated, as they are different for each month?

```lua
local function calculate_dates(date)
    local time = os.time(date)
    local dinfo = os.date("*t", time) -- this normalizes the input to a full date table
    local oneday = 24 * 60 * 60 -- hours * days * seconds
    local oneweek = 7 * oneday
    local df = dateformats
```

I would also like to define a variable to be used for the new months template, I have found these for week, 
```lua
    dates.week = os.date(df.week, time)
    dates.prevweek = os.date(df.week, time - oneweek)
    dates.nextweek = os.date(df.week, time + oneweek)
```
So, could I simply add variables for the month?

```lua
    dates.month = os.date(df.month time)
    dates.prevmonth = os.date(df.month time - onemonth
    dates.nextmonth = os.date(df.week, time + onemonth
```

Also, here just a month variable could be added easily, are there any methods for quarter date calculations too?

```lua
local dateformats = {
    date = "%Y-%m-%d",
    week = "%V",
    isoweek = "%Y-W%V",
}

```

I would also like to know how I could modify other basic variable naming conventions, e.g. for images: 

```lua
    local pngname = "pasted_img_" .. os.date("%Y%m%d%H%M%S") .. ".png"
```

So my most basic question I have in terms of customizing I have for the start:
How would it be possible to modify a locally defined variable like `local pngname`, or `local daymap` in one's configuration lua file`? 

If I just insert such modified lines, e.g.
```lua
local pngname =  os.date("%Y%m%d%H%M%S") .. "Plaatje_Geplakt" .. ".png"
```

, this did not work for me. How does it work? 

My apologies for bad lua knowledge. The ideas in (1. https://github.com/renerocksai/telekasten.nvim/issues/90#issuecomment-1041382294 , 2. https://github.com/renerocksai/telekasten.nvim/issues/94 , 3. https://github.com/renerocksai/telekasten.nvim/issues/99 were very nice!

@lambtho12 

> I think that it would be nice to completely separate the template stuff in a separate file (already discussed https://github.com/renerocksai/telekasten.nvim/issues/90#issuecomment-1041382294). This might facilitate PRs and community improvements. People have lots of interesting ideas, often quite simple to implement (like {{now_month}}, {{last_month}}, {{next_month}}, {{timezone}}).

> In my opinion when it comes to templates and snippets like that, the more we have, the better. 
We could also have stuff not directly related to time, such as {{filename}}, {{pwd}}, {{dir}}, {{host}}, {{user}}, {{some git stuff?}}...

What would be an easy way to create templates and snippets `like that`? I would really like to know how to contribute / where to start for knowing how to, or where to look for as a beginner …

I hope these questions are not too far fetched!

Kind regards :)

**Comment by lambtho12** [comment]
Many things to discuss here... :smile: 

First of all, regarding `hdate`, maybe we could directly use `os.date` without the need for monthmap or daymap at all. See [os.date docs](https://www.lua.org/pil/22.1.html). It seems that it should be possible to just use  `%B` or `%A`. I am not sure, but this might be localized already.

Secondly, regarding the change of `monthmap`, `daymap`, yes simply changing that in `lua/telekasten.lua` would do the trick. But that would only do it on your setup. If the option I discussed just above is not possible, it might be useful to expose `monthmap` and `daymap` as user-configurable options.

Regarding `prevmonth`, `nextmonth` it is not that straightforeward. Do you define a month as 30 days? Or 31? Or 28? or 29? Depending on that, you might get different results than what is expected for some specific days of the year. Depending on if we keep using monthmap or not, we could simply use something like that
```lua
if dinfo.month ~= 1 then
  prevmonth monthmap[dinfo.month-1]
else
  prevmonth = monthmap[12]
end
```

For quarter calculation, it is different. As discussed in #90, some people (business) may decide that the Q1 is not Jan-March, but something else.


----
Now for the modification of stuff and general contribution, here is how I do it.

First of all, you can not just add random stuff in your configuration file and expect it to pass through telekasten. The configuration is in essence just a table with values for some parameters. Adding code in it will not do anything. 

What you need to do to start adding features/improving Telekasten (or any other nvim plugin) is this (at least this is how I do it, I am not the most experienced software dev in the world, there are probably simpler/easier/cleaner ways to proceed):
1. Fork this repository
2. Clone your fork somewhere on your computer and create a new branch for you feature
3. Edit you nvim plugin manager so it loads your local copy instead of `renerocksai/telekasten.nvim` then install the 'new' plugin (i.e. local) and clean the 'old' one (i.e. renerocksai/telekasten)
```lua
  -- With packer, it is simply
  use '~/projects/personal/hover.nvim'
```
4. (Restart nvim, ensure telekasten still works as expected (it should))
5. Start modifying you local copy of the plugin
6. Test, test, test
7. Update docs, readme, etc...
9. Push to github and create a pull request
10. Enjoy seeing it merged and used by ~~millions~~ hundreds of people worldwide thanks to you
---- 

Regarding the more practical aspect, to expose something to the user (i.e. make it modifiable through the config):
1. Add it to the default config table (https://github.com/renerocksai/telekasten.nvim/blob/15f68e7d9fbf050c9c4b60d9ce890c49867f0abf/lua/telekasten.lua#L32-L130)
2. If applicable, ensure it is properly setup in https://github.com/renerocksai/telekasten.nvim/blob/15f68e7d9fbf050c9c4b60d9ce890c49867f0abf/lua/telekasten.lua#L2623-L2711

As you are looking to start with a very basic stuff, you could move the current code for `daymap` and `monthmap` to M.cfg for instance and verify that changing their values in your config will actually make it change.

Thank you very much for wanting to get involved and good luck. We can chat here if you need a bit more help.
